### PR TITLE
[issue-#21] Implement bean factory

### PR DESCRIPTION
**Pull request by MykolaHarasym**
## Checklist before requesting a review
- [x] There are no MR remarks
- [x] Clean and concise code that is easy to read and understand
- [x] Code covered with tests and they are green
- [x] Code API has informative JavaDoc on it
- [ ] Rich logging with detailed messages and correct use of log levels

### CODE REVIEW

File: src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  public BeanFactoryImpl(BeanDefinitionRegistry beanDefinitionRegistry) {
+    this.beanDefinitionRegistry = beanDefinitionRegistry;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Map<String, Object> createBeans() {
+    setRawBeanProcessor(new RawBeanProcessor(beanDefinitionRegistry, rawBeansMap));
```
Comment (vitshch) [inline src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java]: I think it's better to initialise `RawBeanProcessor` in constructor of `BeanFactoryImpl`

File: src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    rawBeanProcessor.process();
+    // TODO: 2/4/2023 create list of bean post processor and process raw beans
+    //beanPostProcessorList.forEach(BeanPostProcessor::process);
```
Comment (vitshch) [inline src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java]: You can have just empty list. It'll not affect your code, and you don't need to comment it

**Review by vitshch — CHANGES_REQUESTED** [review]

File: src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import com.bringframework.factory.RawBeanProcessor;
+import com.bringframework.registry.BeanDefinitionRegistry;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is the main implementation of the {@link BeanFactory} interface.
+ */
```
Comment (marynamelnichuk) [inline src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java:15]: Please add some logging

File: src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+public class BeanFactoryImpl implements BeanFactory {
+  private final Map<String, Object> rawBeansMap = new ConcurrentHashMap<>();
+  private final BeanDefinitionRegistry beanDefinitionRegistry;
+  private final RawBeanProcessor rawBeanProcessor;
+
+  public BeanFactoryImpl(BeanDefinitionRegistry beanDefinitionRegistry) {
+    this.beanDefinitionRegistry = beanDefinitionRegistry;
+    rawBeanProcessor = new RawBeanProcessor(this.beanDefinitionRegistry, rawBeansMap);
```
Comment (marynamelnichuk) [inline src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java]: Please change to `this.rawBeanProcessor = new RawBeanProcessor(this.beanDefinitionRegistry, rawBeansMap);` for consistency

File: src/test/java/com/bringframework/factory/BeanFactoryImplTest.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    beanDefinitionMap.put("str", stringBeanDefinition);
+    when(beanDefinitionRegistry.getAllBeanDefinitions()).thenReturn(beanDefinitionMap);
+
+    BeanFactoryImpl beanFactory = new BeanFactoryImpl(beanDefinitionRegistry);
+
+    //when
+    Map<String, Object> readyBeans = beanFactory.createBeans();
+
+    //then
+    Assertions.assertEquals(1, readyBeans.size());
```
Comment (marynamelnichuk) [inline src/test/java/com/bringframework/factory/BeanFactoryImplTest.java]: please use static import here for consistency

File: src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java (hunk 3 above)
Comment (MykolaHarasym) [reply src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java:15]: RawBeanProcessor and AutowiredBeanPostProcessor contain good logging
I don't think it is necessary here

File: src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    this.rawBeanProcessor = new RawBeanProcessor(this.beanDefinitionRegistry, rawBeansMap);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Map<String, Object> createBeans() {
+    rawBeanProcessor.process();
+    List<BeanPostProcessor> beanPostProcessorList = new ArrayList<>();
```
Comment (vitshch) [inline src/main/java/com/bringframework/factory/impl/BeanFactoryImpl.java]: I think you should have `beanPostProcessorList` as class field and initialise it in the constructor for now

**Review by vitshch — CHANGES_REQUESTED** [review]

**Review by marynamelnichuk — APPROVED** [review]

**Review by vitshch — APPROVED** [review]

**Pull request merged**